clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -67,7 +67,27 @@
 
     public final String logPrefix;
 
-    private final Map<TopicPartition, Integer> sequenceNumbers;
+    // The base sequence of the next batch bound for a given partition.
+    private final Map<TopicPartition, Integer> nextSequence;
```

Comment:
High level comment. There is a lot of per-partition state that we're tracking with separate collections. I wonder if this would be easier to manage with a `PartitionState` object and a single `Map<TopicPartition, PartitionState>`?